x/auth/legacy/v043/store.go

```diff
@@ -0,0 +1,169 @@
+package v043
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+	vestingtypes "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/gogo/protobuf/grpc"
+	"github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+const (
+	delegatorDelegationPath = "/cosmos.staking.v1beta1.Query/DelegatorDelegations"
+	balancesPath            = "/cosmos.bank.v1beta1.Query/AllBalances"
+)
+
+func migrateVestingAccounts(ctx sdk.Context, accounts []types.AccountI, queryServer grpc.Server) ([]types.AccountI, error) {
+	for i := 0; i < len(accounts); i++ {
+		asVesting, ok := accounts[i].(exported.VestingAccount)
+		if !ok {
+			continue
+		}
+
+		account := accounts[i]
+
+		addr := account.GetAddress().String()
+		balance, err := getBalance(
+			ctx,
+			addr,
+			queryServer,
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		delegations, err := getDelegatorDelegationsSum(
+			ctx,
+			addr,
+			queryServer,
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		asVesting, ok = resetVestingDelegatedBalances(asVesting)
+		if !ok {
+			continue
+		}
+
+		// balance before any delegation includes balance of delegation
+		for _, coin := range delegations {
+			balance = balance.Add(coin)
+		}
+
+		asVesting.TrackDelegation(ctx.BlockTime(), balance, delegations)
+
+		accounts[i] = asVesting.(types.AccountI)
+	}
+
+	return accounts, nil
+}
+
+func resetVestingDelegatedBalances(evacct exported.VestingAccount) (exported.VestingAccount, bool) {
+	// reset `DelegatedVesting` and `DelegatedFree` to zero
+	df := sdk.NewCoins()
+	dv := sdk.NewCoins()
+
+	switch vacct := evacct.(type) {
+	case *vestingtypes.ContinuousVestingAccount:
+		vacct.DelegatedVesting = dv
+		vacct.DelegatedFree = df
+		return vacct, true
+	case *vestingtypes.DelayedVestingAccount:
+		vacct.DelegatedVesting = dv
+		vacct.DelegatedFree = df
+		return vacct, true
+	case *vestingtypes.PeriodicVestingAccount:
+		vacct.DelegatedVesting = dv
+		vacct.DelegatedFree = df
+		return vacct, true
+	default:
+		return nil, false
+	}
+}
+
+func getDelegatorDelegationsSum(ctx sdk.Context, address string, queryServer grpc.Server) (sdk.Coins, error) {
```

Comment:
This needs to include unbonding delegations.

`DelegatorDelegations` returns items stored with the `GetDelegationsKey` prefix https://github.com/cosmos/cosmos-sdk/blob/863d7d033f8cf6e22f2cdedc40e02201a17b2821/x/staking/keeper/grpc_query.go#L271

An [`Undelegate`](https://github.com/cosmos/cosmos-sdk/blob/863d7d033f8cf6e22f2cdedc40e02201a17b2821/x/staking/keeper/msg_server.go#L304) message will [`Unbond`](https://github.com/cosmos/cosmos-sdk/blob/863d7d033f8cf6e22f2cdedc40e02201a17b2821/x/staking/keeper/delegation.go#L737) which subtracts from or deletes an entry under that same prefix https://github.com/cosmos/cosmos-sdk/blob/863d7d033f8cf6e22f2cdedc40e02201a17b2821/x/staking/keeper/delegation.go#L112

It then creates an unbonding entry https://github.com/cosmos/cosmos-sdk/blob/863d7d033f8cf6e22f2cdedc40e02201a17b2821/x/staking/keeper/delegation.go#L748-L749

When the unbonding is complete, the originally problematic method in bank keeper is called https://github.com/cosmos/cosmos-sdk/blob/863d7d033f8cf6e22f2cdedc40e02201a17b2821/x/staking/keeper/delegation.go#L782

which eventually calls `TrackUndelegation` https://github.com/cosmos/cosmos-sdk/blob/863d7d033f8cf6e22f2cdedc40e02201a17b2821/x/bank/keeper/keeper.go#L447

So, if this doesn't look at unbonding delegations, there would be a situation where

1. user unbonds with bug in place
2. migration runs, resets vesting account fields, does not account for unbonding delegation
3. unbond completes, `TrackUndelegation` is called for which there was no corresponding `TrackDelegation`.

Reply:
I was just about to push support for unbonding delegations! 

I'll request review from you as soon as I'm done.

Reply:
Is this done?

Reply:
It seams this is still not done.

Reply:
This has been taken into account already in fef2d0075b801f3e448e5ae175b3ab2e70f469c4.